edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs

```diff
@@ -0,0 +1,191 @@
+            // So we should have an authchain already cached for the target device.
+            IEdgeHub edgeHub = await this.edgeHubGetter;
+            Option<string> targetAuthChainOption = await edgeHub.GetAuthChainForIdentity(targetDeviceId);
+            string targetAuthChain = targetAuthChainOption.Expect(() => new UnauthorizedAccessException($"{targetDeviceId} is not a child of this Edge device"));
+
+            // The actor device should be somewhere in the target device's authchain
+            if (!targetAuthChain.Contains(actorDeviceId))
+            {
+                Events.UnauthorizedActor(actorDeviceId, targetDeviceId);
+                result.Status = HttpStatusCode.Unauthorized;
+                return result;
+            }
+
+            // Construct the result
+            IList<ServiceIdentity> identities = await edgeHub.GetDevicesAndModulesInTargetScopeAsync(targetDeviceId);
+            Events.SendingScopeResult(identities);
+            result = MakeResultFromIdentities(identities);
+            result.Status = HttpStatusCode.OK;
+
+            return result;
+        }
+
+        async Task SendResponse(EdgeHubScopeResult result)
+        {
+            this.Response.StatusCode = (int)result.Status;
+
+            if (result.Status == HttpStatusCode.OK)
+            {
+                var resultJsonContent = JsonConvert.SerializeObject(result);
+                var resultUtf8Bytes = Encoding.UTF8.GetBytes(resultJsonContent);
+
+                this.Response.ContentLength = resultUtf8Bytes.Length;
+                this.Response.ContentType = SupportedContentType;
+
+                await this.Response.Body.WriteAsync(resultUtf8Bytes, 0, resultUtf8Bytes.Length);
+            }
+        }
+
+        public static bool ValidateChainAndGetTargetDeviceId(string actorDeviceId, string authChain, out string targetDeviceId)
+        {
+            targetDeviceId = string.Empty;
+            var actorAuthChainIds = authChain.Split(';', StringSplitOptions.RemoveEmptyEntries).ToList();
+
+            // Should have at least 1 element in the chain
+            if (actorAuthChainIds.Count < 1)
+            {
+                return false;
+            }
+
+            // The actor device is the one that sent the request, so the last element
+            // of the auth-chain should always be the actor device's ID
+            if (!actorDeviceId.Equals(actorAuthChainIds.LastOrDefault()))
+            {
+                return false;
+            }
+
+            // The target device ID is the first device ID in the chain
+            if (!actorAuthChainIds[0].Contains('/'))
+            {
+                targetDeviceId = actorAuthChainIds[0];
+            }
+            else
+            {
+                // The first element is a module ID, so the target device
+                // should be the second element in the chain
+                if (actorAuthChainIds.Count < 2 || actorAuthChainIds[1].Contains('/'))
```

Comment:
Should we instead get the "device" that the module belongs to?

Reply:
yes, that's safer than what I'm currnetly doing

---
In reply to: [451661275](https://github.com/Azure/iotedge/pull/3149#discussion_r451661275) [](ancestors = 451661275)